pkg/apis/resource/validation/validation.go

```diff
@@ -0,0 +1,324 @@
+	}
+
+	return allErrs
+}
+
+// ValidateClassUpdate tests if an update to ResourceClass is valid.
+func ValidateClassUpdate(resourceClass, oldClass *resource.ResourceClass) field.ErrorList {
+	allErrs := corevalidation.ValidateObjectMetaUpdate(&resourceClass.ObjectMeta, &oldClass.ObjectMeta, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidateClass(resourceClass)...)
+	return allErrs
+}
+
+// ValidateClaimUpdate tests if an update to ResourceClaim is valid.
+func ValidateClaimUpdate(resourceClaim, oldClaim *resource.ResourceClaim) field.ErrorList {
+	allErrs := corevalidation.ValidateObjectMetaUpdate(&resourceClaim.ObjectMeta, &oldClaim.ObjectMeta, field.NewPath("metadata"))
+	allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(resourceClaim.Spec, oldClaim.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, ValidateClaim(resourceClaim)...)
+	return allErrs
+}
+
+// ValidateClaimStatusUpdate tests if an update to the status of a ResourceClaim is valid.
+func ValidateClaimStatusUpdate(resourceClaim, oldClaim *resource.ResourceClaim) field.ErrorList {
+	allErrs := corevalidation.ValidateObjectMetaUpdate(&resourceClaim.ObjectMeta, &oldClaim.ObjectMeta, field.NewPath("metadata"))
+	fldPath := field.NewPath("status")
+	// The name might not be set yet.
+	if resourceClaim.Status.DriverName != "" {
+		allErrs = append(allErrs, validateResourceDriverName(resourceClaim.Status.DriverName, fldPath.Child("driverName"))...)
+	} else if resourceClaim.Status.Allocation != nil {
+		allErrs = append(allErrs, field.Required(fldPath.Child("driverName"), "must be set when a claim is allocated"))
+	}
+
+	allErrs = append(allErrs, validateAllocationResult(resourceClaim.Status.Allocation, fldPath.Child("allocation"))...)
+	allErrs = append(allErrs, validateSliceIsASet(resourceClaim.Status.ReservedFor, resource.ResourceClaimReservedForMaxSize,
+		validateResourceClaimUserReference, fldPath.Child("reservedFor"))...)
+
+	// Now check for invariants that must be valid for a ResourceClaim.
+	if len(resourceClaim.Status.ReservedFor) > 0 {
+		if resourceClaim.Status.Allocation == nil {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "only allocated claims can be reserved"))
+		} else {
+			if len(resourceClaim.Status.ReservedFor) > 1 {
+				allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "the claim cannot be reserved more than once"))
+			}
+			// Items may be removed from ReservedFor while the claim is meant to be deallocated,
+			// but not added.
+			if resourceClaim.DeletionTimestamp != nil ||
+				resourceClaim.Status.DeallocationRequested {
+				oldSet := sets.New(oldClaim.Status.ReservedFor...)
+				newSet := sets.New(resourceClaim.Status.ReservedFor...)
+				newItems := newSet.Difference(oldSet)
+				if len(newItems) > 0 {
+					allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "new entries may not be added while the claim is meant to be deallocated"))
```

Comment:
Changed.

Reply:
Spoke too fast: because reservedFor is non-empty, the claim is not yet in the process of being deallocated. It can't be, because it's still in use. Deallocation will start once reservedFor is finally empty.

Considering that, "meant to be" is more accurate. Keep it?

Reply:
I try to avoid inferring intention.  "while the claim is meant to be deallocated" is less precise than "while `deallocationRequested` is set".

Reply:
Makes sense, will change it.

Reply:
Done, with "new entries may not be added while `deallocationRequested` or `deletionTimestamp` are set" (it could be one of these two).